src/toil/lib/docker.py

```diff
@@ -1,280 +1,340 @@
-        # When piping, all arguments now get merged into a single string to bash -c.
-        # We try to support spaces in paths by wrapping them all in quotes first.
-        chain_params = [' '.join(p) for p in [list(map(pipes.quote, q)) for q in parameters]]
-        # Use bash's set -eo pipefail to detect and abort on a failure in any command in the chain
-        call = baseDockerCall + ['--entrypoint', '/bin/bash',  tool, '-c',
-                                 'set -eo pipefail && {}'.format(' | '.join(chain_params))]
-    else:
-        call = baseDockerCall + [tool] + parameters
-    _logger.info("Calling docker with " + repr(call))
-
-    params = {}
-    if outfile:
-        params['stdout'] = outfile
-    if checkOutput:
-        callMethod = subprocess.check_output
-    else:
-        callMethod = subprocess.check_call
+    # Ensure the user has passed a valid value for deferParam
+    assert deferParam in (None, FORGO, STOP, RM), \
+        'Please provide a valid value for deferParam.'
 
-    for attempt in retry(predicate=dockerPredicate):
-        with attempt:
-            out = callMethod(call, **params)
+    client = docker.from_env(version='auto')
 
-    _fixPermissions(tool=tool, workDir=workDir)
-    return out
+    if deferParam == STOP:
+        job.defer(_dockerStop, containerName, client)
 
+    if deferParam == FORGO:
+        removeOnExit = False
+    elif deferParam == RM:
+        removeOnExit = True
+        job.defer(_dockerKill, containerName, client)
+    elif removeOnExit is True:
+        job.defer(_dockerKill, containerName, client)
 
-def _dockerKill(containerName, action):
+    try:
+        out = client.containers.run(image=tool,
+                                    command=command,
+                                    working_dir=workDir,
+                                    entrypoint=entrypoint,
+                                    name=containerName,
+                                    detach=detach,
+                                    volumes=
+                                     {workDir: {'bind': dataDir, 'mode': 'rw'}},
+                                    auto_remove=removeOnExit,
+                                    remove=removeOnExit,
+                                    log_config=logDriver,
+                                    user=userPrivilege,
+                                    environment=environment)
+        return out
+    # If the container exits with a non-zero exit code and detach is False.
+    except ContainerError:
+        _logger.error("Docker had non-zero exit.  Check your command: " + \
+                      repr(command))
+        raise
+    except ImageNotFound:
+        _logger.error("Docker image not found.")
+        raise
+    except requests.exceptions.HTTPError as e:
+        _logger.error("The server returned an error.")
+        raise create_api_error_from_http_exception(e)
+    except (OSError, ReferenceError, SystemError,
```

Comment:
Seems redundant to specifically catch and raise for these 6 exceptions then have a blanket except below.

Reply:
Updated.